{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Defining reverse complement function\n",
    "\n",
    "def reverse_complement(DNA):\n",
    "    reverse = DNA.upper()[::-1]\n",
    "    complement = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}\n",
    "    return(''.join([complement[base] for base in reverse]))\n",
    "\n",
    "### Defining function that returns 8 base pair sliding window from input of 15 bp sequence, major allele, minor allele\n",
    "\n",
    "def eightbp_major_minor(seq_15, major, minor):\n",
    "    major_15 = seq_15[0:7] + major + seq_15[8:]\n",
    "    minor_15 = seq_15[0:7] + minor + seq_15[8:]\n",
    "    list_of_major_minor_pairs = []\n",
    "    for i in range(0, 8):\n",
    "        list_of_major_minor_pairs.append([major_15[i:(i+8)].upper(), minor_15[i:(i+8)].upper()])\n",
    "    return(list_of_major_minor_pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "### Importing RBC SNPs\n",
    "\n",
    "import re\n",
    "\n",
    "test = open(\"RBC_SNPs_single_30072018.txt\", \"r\")\n",
    "count = 0\n",
    "rsIDs = {}\n",
    "# list_RBC_rsIDs = []\n",
    "# key_rsIDs = str()\n",
    "checkRepeats = []\n",
    "\n",
    "for readLine in test:\n",
    "    if count == 0:\n",
    "        count += 1\n",
    "        continue\n",
    "    if count > 0:\n",
    "        count += 1\n",
    "        key_rsIDs = (readLine.split())[1]\n",
    "        try:\n",
    "            len((readLine.split())[4]) == 41 & len((readLine.split())[5]) == 41 \n",
    "        except:\n",
    "            print(key_rsIDs + \"is not of length 41\")\n",
    "        if key_rsIDs in rsIDs:\n",
    "            checkRepeats.append(key_rsIDs)\n",
    "        rsIDs[key_rsIDs] = [(readLine.split())[1], ((readLine.split())[4])[13:28].upper(), ((readLine.split())[5])[13:28].upper()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "734\n"
     ]
    }
   ],
   "source": [
    "### Check for Signalling Center SNPs\n",
    "\n",
    "file_Signalling_SNPs = open(\"Signalling_Ctr_SNPs_04092018.txt\", \"r\")\n",
    "list_Signalling_SNPs = []\n",
    "hash_Signalling_SNPs = {}\n",
    "\n",
    "for readLine in file_Signalling_SNPs:\n",
    "    list_Signalling_SNPs.append(readLine.rstrip())\n",
    "\n",
    "for i in list_Signalling_SNPs:\n",
    "    if i in rsIDs:\n",
    "        hash_Signalling_SNPs[i] = rsIDs[i]\n",
    "\n",
    "### Create sliding window of 734 Signalling Centre SNPs\n",
    "\n",
    "hash_SigCtr_SNPs_sliding_window = {}\n",
    "\n",
    "for k, v in hash_Signalling_SNPs.items():\n",
    "    hash_SigCtr_SNPs_sliding_window[k] = []\n",
    "    for i in range(0, 8):\n",
    "        hash_SigCtr_SNPs_sliding_window[k].append([hash_Signalling_SNPs[k][1][i:(i+8)].upper(), hash_Signalling_SNPs[k][2][i:(i+8)].upper()])\n",
    "print(len(hash_SigCtr_SNPs_sliding_window))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Line 67 removed\n",
      "Line 85 removed\n",
      "Line 95 removed\n",
      "Line 120 removed\n",
      "Line 126 removed\n",
      "65\n"
     ]
    }
   ],
   "source": [
    "### Getting all missing Signalling Ctr SNPs from dbSNP, with 5 indels removed - recovered 65 out of 70 absent SNPs\n",
    "### Of these remaining 65 SNPs - picked out the minor allele that had some allele frequency information\n",
    "\n",
    "import re\n",
    "\n",
    "file_vcf_absent_SNPs = open(\"absent_SigCtr_RBC_SNPs_caf.recode.vcf\", \"r\")\n",
    "count = 0\n",
    "hash_vcf_absent_SNPs = {}\n",
    "\n",
    "for readLine in file_vcf_absent_SNPs:\n",
    "    count += 1\n",
    "    if (count >= 58):\n",
    "        list_of_SNP_info = readLine.split()\n",
    "        if len(list_of_SNP_info[3]) > 1:\n",
    "            print(\"Line \" + str(count) + \" removed\")\n",
    "            continue\n",
    "        if len(list_of_SNP_info[4]) == 3:\n",
    "            match_test = re.search(\"CAF=(.*),(.*),(.*)$\", list_of_SNP_info[7])\n",
    "            if match_test.group(2) != \".\":\n",
    "                minor_allele = list_of_SNP_info[4][0]\n",
    "                caf = \"CAF=\" + str(match_test.group(1)) + \",\" + str(match_test.group(2))\n",
    "            if match_test.group(3) != \".\":\n",
    "                minor_allele = list_of_SNP_info[4][2]\n",
    "                caf = \"CAF=\" + str(match_test.group(1)) + \",\" + str(match_test.group(3))\n",
    "            hash_vcf_absent_SNPs[list_of_SNP_info[2]] = (list_of_SNP_info[0], list_of_SNP_info[1],\n",
    "                                                           list_of_SNP_info[2], list_of_SNP_info[3],\n",
    "                                                           minor_allele, caf)\n",
    "            continue\n",
    "        if len(list_of_SNP_info[4]) == 5:\n",
    "            match_test = re.search(\"CAF=(.*),(.*),(.*),(.*)$\", list_of_SNP_info[7])\n",
    "            if match_test.group(2) != \".\":\n",
    "                minor_allele = list_of_SNP_info[4][0]\n",
    "                caf = \"CAF=\" + str(match_test.group(1)) + \",\" + str(match_test.group(2))\n",
    "            if match_test.group(3) != \".\":\n",
    "                minor_allele = list_of_SNP_info[4][2]\n",
    "                caf = \"CAF=\" + str(match_test.group(1)) + \",\" + str(match_test.group(3))\n",
    "            if match_test.group(4) != \".\":\n",
    "                minor_allele = list_of_SNP_info[4][4]\n",
    "                caf = \"CAF=\" + str(match_test.group(1)) + \",\" + str(match_test.group(4))\n",
    "            hash_vcf_absent_SNPs[list_of_SNP_info[2]] = (list_of_SNP_info[0], list_of_SNP_info[1],\n",
    "                                                           list_of_SNP_info[2], list_of_SNP_info[3],\n",
    "                                                           minor_allele, caf)\n",
    "            continue\n",
    "        else:\n",
    "            hash_vcf_absent_SNPs[list_of_SNP_info[2]] = (list_of_SNP_info[0], list_of_SNP_info[1],\n",
    "                                                           list_of_SNP_info[2], list_of_SNP_info[3],\n",
    "                                                           list_of_SNP_info[4], list_of_SNP_info[7])\n",
    "\n",
    "print(len(hash_vcf_absent_SNPs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "### Using Biopython SeqIO module to obtain 15-bp window with each SNP at the centre - \n",
    "### for obtaining sliding window of 8-mers for SNPs\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "list_chr = []\n",
    "\n",
    "for i in range(1, 23):\n",
    "    list_chr.append(i)\n",
    "list_chr.append(\"X\")\n",
    "list_chr.append(\"Y\")\n",
    "\n",
    "hash_RBC_SigCtr_SNPs_8mer_major_minor = {}\n",
    "\n",
    "for str_chr in list_chr:\n",
    "    str_file_name_chr = \"chr\" + str(str_chr) + \".fa\"\n",
    "    for seq_record in SeqIO.parse(str_file_name_chr, \"fasta\"):\n",
    "        for k, v in hash_vcf_absent_SNPs.items():\n",
    "            if v[0] == str(str_chr):\n",
    "                chrpos = int(v[1])\n",
    "                seq_15_current = str(seq_record.seq[(chrpos-8):(chrpos+7)]).upper()\n",
    "                if len(v[4]) == 1:\n",
    "                    hash_RBC_SigCtr_SNPs_8mer_major_minor[k] = eightbp_major_minor(seq_15_current, v[3], v[4])\n",
    "                else:\n",
    "                    print(hash_vcf_absent_SNPs[k])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "799\n"
     ]
    }
   ],
   "source": [
    "### Combining the existing list of Signalling Ctr RBC SNPs, with those that were absent in the RBC SNPs list - total 799\n",
    "\n",
    "hash_combined_SigCtr_SNPs = {}\n",
    "\n",
    "for k, v in hash_RBC_SigCtr_SNPs_8mer_major_minor.items():\n",
    "    hash_combined_SigCtr_SNPs[k] = hash_RBC_SigCtr_SNPs_8mer_major_minor[k]\n",
    "\n",
    "for k, v in hash_SigCtr_SNPs_sliding_window.items():\n",
    "    hash_combined_SigCtr_SNPs[k] = hash_SigCtr_SNPs_sliding_window[k]\n",
    "\n",
    "print(len(hash_combined_SigCtr_SNPs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "### Extracting relevant information regarding rsIDs of interest from Framingham Heart Study cis-eQTL Excel spreadsheet\n",
    "\n",
    "import pandas as pd, os, re\n",
    "\n",
    "hash_rsIDs_with_ciseQTL_gene = {}\n",
    "hash_rsIDs_with_ciseQTL_exon = {}\n",
    "\n",
    "pd_df_Framingham = pd.read_excel(\"FHS_SNP_SignalingCenters_new.xlsx\", sheetname = 1, header = None, skiprows = 4)\n",
    "list_rsID = pd_df_Framingham[0].tolist()\n",
    "list_SNP_Fx_Allele = pd_df_Framingham[6].tolist()\n",
    "list_SNP_Non_Fx_Allele = pd_df_Framingham[7].tolist()\n",
    "list_T = pd_df_Framingham[22].tolist()\n",
    "list_log10FDR = pd_df_Framingham[24].tolist()\n",
    "\n",
    "for i in range(0, 873):\n",
    "    hash_rsIDs_with_ciseQTL_gene[list_rsID[i]] = (list_SNP_Fx_Allele[i], list_SNP_Non_Fx_Allele[i], list_T[i], list_log10FDR[i])\n",
    "\n",
    "exon_pd_df_Framingham = pd.read_excel(\"FHS_SNP_SignalingCenters_new.xlsx\", sheetname = 3, header = None, skiprows = 4)\n",
    "exon_list_rsID = exon_pd_df_Framingham[0].tolist()\n",
    "exon_list_SNP_Fx_Allele = exon_pd_df_Framingham[6].tolist()\n",
    "exon_list_SNP_Non_Fx_Allele = exon_pd_df_Framingham[7].tolist()\n",
    "exon_list_T = exon_pd_df_Framingham[22].tolist()\n",
    "exon_list_log10FDR = exon_pd_df_Framingham[24].tolist()\n",
    "\n",
    "for i in range(0, 873):\n",
    "    hash_rsIDs_with_ciseQTL_exon[exon_list_rsID[i]] = (exon_list_SNP_Fx_Allele[i], exon_list_SNP_Non_Fx_Allele[i], exon_list_T[i], exon_list_log10FDR[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "### Sliding window of 3,263 RBC SNPs\n",
    "\n",
    "hash_RBC_SNPs_sliding_window = {}\n",
    "\n",
    "for k, v in rsIDs.items():\n",
    "    hash_RBC_SNPs_sliding_window[k] = []\n",
    "    for i in range(0, 8):\n",
    "        hash_RBC_SNPs_sliding_window[k].append([rsIDs[k][1][i:(i+8)].upper(), rsIDs[k][2][i:(i+8)].upper()])\n",
    "\n",
    "### Generating dictionary versions of individual PBM datasets of interest for evaluating perturbation of TF binding\n",
    "### The reverse_complement() function is needed to ensure that all forward and reverse 8-mer orientations are present.\n",
    "\n",
    "hash_all_PBMs_with_Escores = {}\n",
    "\n",
    "for file in os.listdir(\"Individual_PBMs_Escores/\"):\n",
    "    if re.search(\"_Escore.txt\", file):\n",
    "        m = re.match(\"(.*)_Escore\\.txt\", file)\n",
    "        str_PBM_name = (m.group(1))\n",
    "        PBM_file = open(\"Individual_PBMs_Escores/\" + file, \"r\")\n",
    "        hash_all_PBMs_with_Escores.setdefault(str_PBM_name, {})\n",
    "        for readLine in PBM_file:\n",
    "            key = readLine.rstrip(\"\\n\").split(\"\\t\")[0]\n",
    "            value = readLine.rstrip(\"\\n\").split(\"\\t\")[1]\n",
    "            key_RC = reverse_complement(key)\n",
    "            hash_all_PBMs_with_Escores[str_PBM_name][key] = value\n",
    "            hash_all_PBMs_with_Escores[str_PBM_name][key_RC] = value\n",
    "        PBM_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Egr1_2580_contig8mers': ('Egr1_2580_contig8mers', 26104, 466, 466, 439), 'EGR2_REF_R1_8mers': ('EGR2_REF_R1_8mers', 26104, 500, 504, 442), 'Gata3_1024_contig8mers': ('Gata3_1024_contig8mers', 26104, 425, 344, 326), 'Gata4_8mers_11111111': ('Gata4_8mers_11111111', 26104, 201, 145, 172), 'Gata5_3768_contig8mers': ('Gata5_3768_contig8mers', 26104, 291, 202, 209), 'Gata6_3769_contig8mers': ('Gata6_3769_contig8mers', 26104, 365, 267, 235), 'GATA_average': ('GATA_average', 26104, 276, 181, 184), 'Gli1_v015681_8mers_11111111': ('Gli1_v015681_8mers_11111111', 26104, 455, 444, 487), 'Gli2_v015681_8mers_11111111': ('Gli2_v015681_8mers_11111111', 26104, 472, 450, 504), 'Gli3_v015681_8mers_11111111': ('Gli3_v015681_8mers_11111111', 26104, 506, 499, 517), 'HES1_R1': ('HES1_R1', 26104, 190, 148, 196), 'HES5_R1': ('HES5_R1', 26104, 364, 321, 392), 'HES7_R1': ('HES7_R1', 26104, 318, 287, 328), 'KLF1_REF_R2_8mers': ('KLF1_REF_R2_8mers', 26104, 224, 182, 226), 'Rara_1051_contig8mers': ('Rara_1051_contig8mers', 26104, 231, 265, 271), 'Rarb_8mers_11111111': ('Rarb_8mers_11111111', 26104, 133, 162, 138), 'Rarg_8mers_11111111': ('Rarg_8mers_11111111', 26104, 158, 206, 177), 'Rxra_1035_contig8mers': ('Rxra_1035_contig8mers', 26104, 259, 288, 284), 'Rxrg_8mers_11111111': ('Rxrg_8mers_11111111', 26104, 141, 183, 180), 'Sfpi1_1034_contig8mers': ('Sfpi1_1034_contig8mers', 26104, 380, 460, 360), 'Smad3_3805_contig8mers': ('Smad3_3805_contig8mers', 26104, 332, 354, 306), 'TCF4_R1': ('TCF4_R1', 26104, 142, 145, 140)}\n"
     ]
    }
   ],
   "source": [
    "### Analysing 3,263 RBC SNPs for SNPs that perturb TF binding (based on individual PBM datasets)\n",
    "\n",
    "hash_PBM_statistics = {}\n",
    "list_rsIDs_PBMs_of_interest = []\n",
    "\n",
    "for str_PBM_name in hash_all_PBMs_with_Escores.keys():\n",
    "\n",
    "    count_loss_binding = 0\n",
    "    count_surprise_no_loss_binding = 0\n",
    "    count_gain_binding = 0\n",
    "    count_surprise_no_gain_binding = 0\n",
    "    testcount = 0\n",
    "    \n",
    "    for k, v in hash_RBC_SNPs_sliding_window.items():        \n",
    "        testcount8mer = 0\n",
    "        for i in range(0, 8):\n",
    "            allele1_key = v[i][0]\n",
    "            allele2_key = v[i][1]\n",
    "            if allele1_key not in hash_all_PBMs_with_Escores[str_PBM_name]:\n",
    "                print(allele1_key)\n",
    "                continue\n",
    "            if allele2_key not in hash_all_PBMs_with_Escores[str_PBM_name]:\n",
    "                print(allele2_key)\n",
    "                continue\n",
    "            if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) > 0.35) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) < 0.3):\n",
    "                count_loss_binding += 1\n",
    "                testcount8mer += 1\n",
    "#                print(k + \"\\t\" + str_PBM_name + \"\\t\" + allele1_key + \"\\t\" + allele2_key + \"\\t\" + str(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) + \"\\t\" + str(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]))\n",
    "                list_rsIDs_PBMs_of_interest.append([k, str_PBM_name])\n",
    "            if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) > 0.3) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) <= float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key])):\n",
    "                count_surprise_no_loss_binding += 1\n",
    "            if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) > 0.35) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) < 0.3):\n",
    "                count_gain_binding += 1\n",
    "                testcount8mer += 1\n",
    "#                print(k + \"\\t\" + str_PBM_name + \"\\t\" + allele1_key + \"\\t\" + allele2_key + \"\\t\" + str(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) + \"\\t\" + str(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]))\n",
    "                list_rsIDs_PBMs_of_interest.append([k, str_PBM_name])\n",
    "            if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) > 0.3) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) <= float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key])):\n",
    "                count_surprise_no_gain_binding += 1\n",
    "        if testcount8mer > 0:\n",
    "            testcount += 1\n",
    "\n",
    "    hash_PBM_statistics[str_PBM_name] = (str_PBM_name, len(hash_RBC_SNPs_sliding_window)*8, count_loss_binding, count_gain_binding, testcount)\n",
    "#    print(hash_PBM_statistics[str_PBM_name])\n",
    "print(hash_PBM_statistics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Egr1_2580_contig8mers 26104 466 466 439 932\n",
      "EGR2_REF_R1_8mers 26104 500 504 442 1004\n",
      "Gata3_1024_contig8mers 26104 425 344 326 769\n",
      "Gata4_8mers_11111111 26104 201 145 172 346\n",
      "Gata5_3768_contig8mers 26104 291 202 209 493\n",
      "Gata6_3769_contig8mers 26104 365 267 235 632\n",
      "GATA_average 26104 276 181 184 457\n",
      "Gli1_v015681_8mers_11111111 26104 455 444 487 899\n",
      "Gli2_v015681_8mers_11111111 26104 472 450 504 922\n",
      "Gli3_v015681_8mers_11111111 26104 506 499 517 1005\n",
      "HES1_R1 26104 190 148 196 338\n",
      "HES5_R1 26104 364 321 392 685\n",
      "HES7_R1 26104 318 287 328 605\n",
      "KLF1_REF_R2_8mers 26104 224 182 226 406\n",
      "Rara_1051_contig8mers 26104 231 265 271 496\n",
      "Rarb_8mers_11111111 26104 133 162 138 295\n",
      "Rarg_8mers_11111111 26104 158 206 177 364\n",
      "Rxra_1035_contig8mers 26104 259 288 284 547\n",
      "Rxrg_8mers_11111111 26104 141 183 180 324\n",
      "Sfpi1_1034_contig8mers 26104 380 460 360 840\n",
      "Smad3_3805_contig8mers 26104 332 354 306 686\n",
      "TCF4_R1 26104 142 145 140 287\n"
     ]
    }
   ],
   "source": [
    "### Using individual PBMs to scan 3263 RBC SNPs - what types of binding changes do we see? \n",
    "### Criteria: E-score > 0.35 for one allele and E-score < 0.3 for other allele\n",
    "\n",
    "hash_PBM_statistics = {}\n",
    "\n",
    "for str_PBM_name in hash_all_PBMs_with_Escores.keys():\n",
    "\n",
    "    count_loss_binding = 0\n",
    "    count_surprise_no_loss_binding = 0\n",
    "    count_gain_binding = 0\n",
    "    count_surprise_no_gain_binding = 0\n",
    "    list_eightmer_change = []\n",
    "    \n",
    "    for k, v in hash_RBC_SNPs_sliding_window.items():\n",
    "        eight8mer_change = 0\n",
    "        for i in range(0, 8):\n",
    "            allele1_key = v[i][0]\n",
    "            allele2_key = v[i][1]\n",
    "            if allele1_key not in hash_all_PBMs_with_Escores[str_PBM_name]:\n",
    "                print(allele1_key)\n",
    "                continue\n",
    "            if allele2_key not in hash_all_PBMs_with_Escores[str_PBM_name]:\n",
    "                print(allele2_key)\n",
    "                continue\n",
    "            if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) > 0.35) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) < 0.3):\n",
    "                count_loss_binding += 1\n",
    "                eight8mer_change += 1\n",
    "            if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) > 0.3) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) <= float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key])):\n",
    "                count_surprise_no_loss_binding += 1\n",
    "            if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) > 0.35) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) < 0.3):\n",
    "                count_gain_binding += 1\n",
    "                eight8mer_change += 1\n",
    "            if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) > 0.3) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) <= float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key])):\n",
    "                count_surprise_no_gain_binding += 1\n",
    "        list_eightmer_change.append(eight8mer_change)\n",
    "\n",
    "#    testing = open(str_PBM_name + \".txt\", \"w\")\n",
    "#    for item in list_eightmer_change:\n",
    "#        testing.write(\"%s\\n\" % item)\n",
    "    hash_PBM_statistics[str_PBM_name] = (str_PBM_name, len(hash_RBC_SNPs_sliding_window)*8, count_loss_binding, count_gain_binding, list_eightmer_change, sum(list_eightmer_change))\n",
    "    count_item_eightmer = 0\n",
    "    for item_eightmer in list_eightmer_change:\n",
    "        if item_eightmer > 0:\n",
    "            count_item_eightmer += 1\n",
    "    print(str_PBM_name, len(hash_RBC_SNPs_sliding_window)*8, count_loss_binding, count_gain_binding, count_item_eightmer, sum(list_eightmer_change))\n",
    "#    testing.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('GATA_average', 26104, 276, 181)\n",
      "('GATA_average', 6392, 91, 67)\n"
     ]
    }
   ],
   "source": [
    "### Getting average GATA PBM dataset to represent GATA3, GATA4, GATA5 and GATA6 datasets\n",
    "\n",
    "hash_all_PBMs_with_Escores[\"GATA_average\"] = {}\n",
    "#file_GATA_average = open(\"GATA_average_Escore.txt\", \"w\")\n",
    "\n",
    "for i in hash_all_PBMs_with_Escores[\"Gata3_1024_contig8mers\"]:\n",
    "    Escore_Gata3 = float(hash_all_PBMs_with_Escores[\"Gata3_1024_contig8mers\"][i])\n",
    "    Escore_Gata4 = float(hash_all_PBMs_with_Escores[\"Gata4_8mers_11111111\"][i])\n",
    "    Escore_Gata5 = float(hash_all_PBMs_with_Escores[\"Gata5_3768_contig8mers\"][i])\n",
    "    Escore_Gata6 = float(hash_all_PBMs_with_Escores[\"Gata6_3769_contig8mers\"][i])\n",
    "    average_Escore_Gata = (Escore_Gata3 + Escore_Gata4 + Escore_Gata5 + Escore_Gata6)/4\n",
    "    hash_all_PBMs_with_Escores[\"GATA_average\"][i] = average_Escore_Gata\n",
    "#    file_GATA_average.write(i + \"\\t\" + str(average_Escore_Gata) + \"\\n\")\n",
    "#file_GATA_average.close()\n",
    "\n",
    "### GATA average for 3263 RBC SNPs\n",
    "\n",
    "str_PBM_name = \"GATA_average\"\n",
    "\n",
    "count_loss_binding = 0\n",
    "count_surprise_no_loss_binding = 0\n",
    "count_gain_binding = 0\n",
    "count_surprise_no_gain_binding = 0\n",
    "\n",
    "for k, v in hash_RBC_SNPs_sliding_window.items():\n",
    "    for i in range(0, 8):\n",
    "        allele1_key = v[i][0]\n",
    "        allele2_key = v[i][1]\n",
    "        if allele1_key not in hash_all_PBMs_with_Escores[str_PBM_name]:\n",
    "            print(allele1_key)\n",
    "            continue\n",
    "        if allele2_key not in hash_all_PBMs_with_Escores[str_PBM_name]:\n",
    "            print(allele2_key)\n",
    "            continue\n",
    "        if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) > 0.35) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) < 0.3):\n",
    "            count_loss_binding += 1\n",
    "        if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) > 0.3) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) <= float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key])):\n",
    "            count_surprise_no_loss_binding += 1\n",
    "        if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) > 0.35) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) < 0.3):\n",
    "            count_gain_binding += 1\n",
    "        if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) > 0.3) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) <= float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key])):\n",
    "            count_surprise_no_gain_binding += 1\n",
    "\n",
    "hash_PBM_statistics[str_PBM_name] = (str_PBM_name, len(hash_RBC_SNPs_sliding_window)*8, count_loss_binding, count_gain_binding)\n",
    "print(hash_PBM_statistics[str_PBM_name])\n",
    "    \n",
    "### GATA average for 799 Signalling Centre SNPs\n",
    "\n",
    "str_PBM_name = \"GATA_average\"\n",
    "\n",
    "count_loss_binding = 0\n",
    "count_surprise_no_loss_binding = 0\n",
    "count_gain_binding = 0\n",
    "count_surprise_no_gain_binding = 0\n",
    "\n",
    "for k, v in hash_combined_SigCtr_SNPs.items():\n",
    "    for i in range(0, 8):\n",
    "        allele1_key = v[i][0]\n",
    "        allele2_key = v[i][1]\n",
    "        if allele1_key not in hash_all_PBMs_with_Escores[str_PBM_name]:\n",
    "            print(allele1_key)\n",
    "            continue\n",
    "        if allele2_key not in hash_all_PBMs_with_Escores[str_PBM_name]:\n",
    "            print(allele2_key)\n",
    "            continue\n",
    "        if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) > 0.35) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) < 0.3):\n",
    "            count_loss_binding += 1\n",
    "        if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) > 0.3) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) <= float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key])):\n",
    "            count_surprise_no_loss_binding += 1\n",
    "        if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) > 0.35) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key]) < 0.3):\n",
    "            count_gain_binding += 1\n",
    "        if (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) > 0.3) & (float(hash_all_PBMs_with_Escores[str_PBM_name][allele2_key]) <= float(hash_all_PBMs_with_Escores[str_PBM_name][allele1_key])):\n",
    "            count_surprise_no_gain_binding += 1\n",
    "\n",
    "hash_PBM_statistics[str_PBM_name] = (str_PBM_name, len(hash_combined_SigCtr_SNPs)*8, count_loss_binding, count_gain_binding)\n",
    "print(hash_PBM_statistics[str_PBM_name])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['ATAAGTGC', 'ATAAGTGT'], ['TAAGTGCC', 'TAAGTGTC'], ['AAGTGCCT', 'AAGTGTCT'], ['AGTGCCTG', 'AGTGTCTG'], ['GTGCCTGG', 'GTGTCTGG'], ['TGCCTGGC', 'TGTCTGGC'], ['GCCTGGCT', 'GTCTGGCT'], ['CCTGGCTC', 'TCTGGCTC']]\n",
      "Gata3_1024_contig8mers\n",
      "('0.247506381638872', '0.066534966836071')\n",
      "('-0.130541678029526', '-0.127687491037713')\n",
      "('-0.140163464295956', '-0.206467939117781')\n",
      "('-0.089551168682185', '-0.065600802982507')\n",
      "('-0.025000265536543', '-0.100583090379009')\n",
      "('-0.192403893641766', '-0.141880860077635')\n",
      "('-0.324310832842114', '-0.212191649183182')\n",
      "('-0.283582847447907', '-0.11384335154827')\n",
      "Gata4_8mers_11111111\n",
      "('0.26568', '0.10543')\n",
      "('-0.07743', '0.00538')\n",
      "('-0.24084', '-0.08602')\n",
      "('-0.00667', '0.01762')\n",
      "('0.07895', '0.08521')\n",
      "('-0.07477', '0.0102')\n",
      "('0.00253', '-0.02308')\n",
      "('-0.08513', '0.04092')\n",
      "Gata5_3768_contig8mers\n",
      "('0.111846249015859', '0.093296365345699')\n",
      "('-0.077806208525475', '-0.109344074198021')\n",
      "('-0.116419947841144', '-0.246190332021395')\n",
      "('0.013476180172201', '-0.029739046374641')\n",
      "('0.035953492571878', '-0.055428079736125')\n",
      "('-0.1786122132458', '-0.163377297819556')\n",
      "('-0.257737516664808', '-0.12345449147297')\n",
      "('-0.177454729501743', '-0.085406143272198')\n",
      "Gata6_3769_contig8mers\n",
      "('0.293397955626525', '0.097818987079609')\n",
      "('-0.141362868332345', '-0.047122508484298')\n",
      "('-0.154890120978449', '-0.209086129433133')\n",
      "('-0.066217776648495', '-0.050818383625955')\n",
      "('0.023802695726986', '-0.016056227330901')\n",
      "('-0.134450657702581', '0.0129421709237')\n",
      "('-0.263537733594609', '-0.17119727911692')\n",
      "('-0.255122178865811', '-0.08251419224989')\n",
      "GATA_average\n",
      "(0.22960764657031402, 0.09077007981534475)\n",
      "(-0.10678518872183648, -0.06969351843000801)\n",
      "(-0.16307838327888724, -0.18694110014307724)\n",
      "(-0.03724069128961975, -0.032134558245775754)\n",
      "(0.028426480690580253, -0.021714349361508754)\n",
      "(-0.14505919114753676, -0.07052899674337276)\n",
      "(-0.21076402077538275, -0.132480854943268)\n",
      "(-0.20032243895386526, -0.06021092176758949)\n",
      "Smad3_3805_contig8mers\n",
      "('-0.145696452498195', '0.132073118964236')\n",
      "('0.034353053542334', '0.107581908491738')\n",
      "('-0.22424662113759', '0.387373906214347')\n",
      "('-0.098270080588806', '0.448946941114666')\n",
      "('-0.044685355875002', '0.478523294067707')\n",
      "('-0.038880929640242', '0.490302195454248')\n",
      "('0.024702854717181', '0.447096955819033')\n",
      "('-0.052145102346368', '0.201668789132973')\n"
     ]
    }
   ],
   "source": [
    "print(hash_RBC_SNPs_sliding_window[\"rs737092\"])\n",
    "\n",
    "tuple_GATA = (\"Gata3_1024_contig8mers\", \"Gata4_8mers_11111111\", \"Gata5_3768_contig8mers\", \"Gata6_3769_contig8mers\", \"GATA_average\", \"Smad3_3805_contig8mers\")\n",
    "    \n",
    "for item in tuple_GATA:\n",
    "    print(item)\n",
    "#    testing = open(item + \"_rs737092\" + \".txt\", \"w\")\n",
    "    for pair in hash_RBC_SNPs_sliding_window[\"rs737092\"]:\n",
    "        print((hash_all_PBMs_with_Escores[item][pair[0]], hash_all_PBMs_with_Escores[item][pair[1]]))\n",
    "#        testing.write(pair[0] + \"\\t\" + pair[1] + \"\\t\" + str(hash_all_PBMs_with_Escores[item][pair[0]]) + \"\\t\" + str(hash_all_PBMs_with_Escores[item][pair[1]]) + \"\\n\")\n",
    "#    testing.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "184\n",
      "457\n",
      "115\n",
      "dict_keys(['57412|100528007', 220972, 93550, 51592, 54855, 91181, '3310|3311', 746, 51430, 143686, 81846, 7832, 493, 25778, 388730, 80342, '89894|728543', 9761, 8638, 144406, '6007|6006', 79947, 10162, 28977, 9731, '23499|643314', 4140, 6513, '283635|55012', 1073, 8503, '91612|2877', 10158, 6886, 57143, 374618, '10260|79982|646358', '83440|100287559', '4123|79661|693216', '55466|80349', 55466, 8131, 3681, 146198, 8140, 11216, 92521, 4642, 55876, 6521, '9842|440456', '8742|407977|8741', '9489|8632', 147339, 1777, 4218, 80726, 84902, 51588, '112398|100529264|100652943', 92799, 23645, '81502|100101490', 171023, '100127886|29053', 10018, 6453, 23076, 7327, 64114, 1593, 150275, 8897, 266629, 23761, 25793, 80895, 267020, 3030, 53335, 10678, 5341, 5922, 7037, 8850, 11180, 7318, '2996|2994|2993', 79667, 132660, 345275, 4299, 10826, '6610|285755', 91749, 10767, 64577, 3720, 29116, 9444, '8366|554313|121504|8370|8368|8367|8365|8364|8363|8362|8361|8360|8359|8294', '8364|554313|121504|8370|8368|8367|8366|8365|8363|8362|8361|8360|8359|8294', 11118, '100188893|29964', 896, 2050, 7328, 26266, 83637, 63979, 4824, 286, 760, 90459, 10171])\n",
      "{2050, '100188893|29964', 171023, 10767, 144406, 9761, 4642, 23076, 91181, 55876, 54855, '6007|6006', 79947, 53335, 23645, 3681, 91749, 64114, 4218, 388730, 92799, 3720, 8850, 7318, 7832, 26266, 7327, 55466, 25778, '283635|55012', 345275, '3310|3311', '9842|440456', 8897, 25793, 4299, 23761, 4824, 5341, 9444, 6886, 1777, 760, 150275, 267020, 146198, '8366|554313|121504|8370|8368|8367|8365|8364|8363|8362|8361|8360|8359|8294', 286, 5922, 220972, 79667, '10260|79982|646358', 6453, '55466|80349', '81502|100101490', 374618, 90459, '23499|643314', 92521, 11118, 7037, 896, 51588, 266629, 51592, '6610|285755', 147339, 84902, '8742|407977|8741', 11180, '4123|79661|693216', 10158, 10162, 10678, 10171, 29116, 8638, 8131, 8140, 11216, 80342, 3030, 63979, 493, '57412|100528007', 80895}\n",
      "86\n"
     ]
    }
   ],
   "source": [
    "### Check Framingham gene dataset for T > 0 or < 0 - indicative of cis-eQTL\n",
    "### Identify number of genes that meet this criteria - 115 such entries, grouped by Entrez ID\n",
    "### For each of the Entrez IDs, created a list of all the rsIDs that are cis-eQTLs for the respective Entrez IDs\n",
    "### Created a list of genes (list_of_genes_with_ciseQTL_perturbing_TF_binding) to contain those that were affected by one of the STF rsIDs:\n",
    "### For every Entrez ID, if one of the 1557 rsIDs_with_altered_binding is present in the list of cis-eQTLs rsIDs of an Entrez ID\n",
    "### Save that Entrez ID in list_of_genes_with_ciseQTL_perturbing_TF_binding\n",
    "### We see 86 such unique Entrez IDs in this list.\n",
    "\n",
    "test_df_Framingham_filtered_gene = pd_df_Framingham.loc[(pd_df_Framingham[22] > 0) | (pd_df_Framingham[22] < 0), : ]\n",
    "rsIDs_test_df_Framingham_filtered_gene = test_df_Framingham_filtered_gene[0].tolist()\n",
    "genes_test_df_Framingham_filtered_gene = test_df_Framingham_filtered_gene[18].tolist()\n",
    "\n",
    "testcount = 0\n",
    "for i in rsIDs_with_altered_binding:\n",
    "    if i in rsIDs_test_df_Framingham_filtered_gene:\n",
    "        testcount += 1\n",
    "        #print(i)\n",
    "print(testcount) \n",
    "print(len(genes_test_df_Framingham_filtered_gene))\n",
    "print(len(set(genes_test_df_Framingham_filtered_gene)))\n",
    "\n",
    "hash_genes_Framingham = {}\n",
    "for i in range(0, len(genes_test_df_Framingham_filtered_gene)):\n",
    "    hash_genes_Framingham.setdefault(genes_test_df_Framingham_filtered_gene[i], [])\n",
    "    (hash_genes_Framingham[genes_test_df_Framingham_filtered_gene[i]]).append(rsIDs_test_df_Framingham_filtered_gene[i])\n",
    "print(hash_genes_Framingham.keys())\n",
    "\n",
    "list_of_genes_with_ciseQTL_perturbing_TF_binding = []\n",
    "for k, v in hash_genes_Framingham.items():\n",
    "    for i in rsIDs_with_altered_binding:\n",
    "        if i in v:\n",
    "            list_of_genes_with_ciseQTL_perturbing_TF_binding.append(k)\n",
    "print(set(list_of_genes_with_ciseQTL_perturbing_TF_binding))\n",
    "print(len(set(list_of_genes_with_ciseQTL_perturbing_TF_binding)))\n",
    "for i in list_of_genes_with_ciseQTL_perturbing_TF_binding:\n",
    "    if i not in hash_genes_Framingham.keys():\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "235\n",
      "593\n",
      "144\n",
      "dict_keys([22978, 57584, 220972, 93550, 51592, 7812, 54855, nan, 9898, 2213, 746, 92346, 23052, 143686, 81846, 7832, 493, 5929, 25778, 80342, 6311, 9761, 51367, 5715, '6007|6006', 79947, 10162, 11163, '1997|11193', 57470, '23499|643314', 9895, 4140, 6513, 11177, 8503, 4149, '23503|145226', 91748, 6886, 9517, 374618, '10260|79982|646358', '83440|100287559', 10066, '55466|80349', 55466, 8131, '9726|79759', 146198, '497190|283971|348174', 8140, 11216, 92521, 4642, 55876, 7343, '51326|100506084|474170', '100533955|1973', '9489|8632', 147339, 1777, 811, 4218, '54858|23770', 84902, '1938|26812', '112398|100529264|100652943', 23645, 22974, '81502|100101490', 5326, 5335, '5770|140876', '55544|100291105', 10018, 6453, 23076, 7327, 10109, 64114, 1593, 150275, 84164, '284904|730005', 23761, 25793, 4627, 7263, 80895, '91695|253190|27341', 53947, 3030, 53335, '130872|9736', '3921|574040|6044|388524', 10678, 5341, 5922, 23228, 7037, '200958|440993', '116211|255758|100534611', 8850, 51447, 387, 275, 4286, '2996|2994|2993', 79667, 170825, 132660, 345275, 4299, 10826, 10788, 9896, 27244, 91749, 64577, 10767, 10370, 3720, 2766, 10475, '8366|554313|121504|8370|8368|8367|8365|8364|8363|8362|8361|8360|8359|8294', '8364|554313|121504|8370|8368|8367|8366|8365|8363|8362|8361|8360|8359|8294', 10473, 7718, '100188893|29964', 896, 6239, 2050, 7328, 26266, 83637, 10320, 63979, 1021, 51312, 286, 760, 90459, 10171])\n",
      "{nan, 2050, '100188893|29964', 10767, '9726|79759', 4627, 9761, 4642, 23076, '130872|9736', 64577, 55876, 54855, '6007|6006', 79947, '51326|100506084|474170', 10320, 5715, 23645, 7263, 6239, 91748, 91749, 27244, 51312, 64114, 4218, '3921|574040|6044|388524', 10370, 7812, 3720, 8850, 7832, 26266, 7327, 2213, 51367, 9896, 9898, 55466, 25778, 53947, 345275, 25793, 84164, '497190|283971|348174', 4299, 5326, 2766, 23761, 5341, 6886, 10473, 10475, 57584, 1777, '116211|255758|100534611', 51447, 760, 150275, 275, 146198, 286, 10018, 5922, '200958|440993', 5929, 811, 220972, '284904|730005', 79667, '10260|79982|646358', 6453, '112398|100529264|100652943', '55544|100291105', 10066, '55466|80349', '81502|100101490', 374618, 90459, '23499|643314', 92521, '1938|26812', '100533955|1973', 7037, 10109, 896, 387, 51592, 147339, 84902, 11177, 10162, '23503|145226', 10678, 10171, 22974, 22978, 8131, '91695|253190|27341', 8140, 11216, 80342, 3030, 63979, 493, 80895}\n",
      "107\n"
     ]
    }
   ],
   "source": [
    "### Apply similar workflow from Framingham gene dataset to Framingham exon dataset\n",
    "### See 107 unique entries out of 144 total entries, but need to account for the \"NaN\" entries\n",
    "\n",
    "test_df_Framingham_filtered_exon = exon_pd_df_Framingham.loc[(exon_pd_df_Framingham[22] > 0) | (exon_pd_df_Framingham[22] < 0), : ]\n",
    "rsIDs_test_df_Framingham_filtered_exon = test_df_Framingham_filtered_exon[0].tolist()\n",
    "genes_test_df_Framingham_filtered_exon = test_df_Framingham_filtered_exon[18].tolist()\n",
    "\n",
    "testcount = 0\n",
    "for i in rsIDs_with_altered_binding:\n",
    "    if i in rsIDs_test_df_Framingham_filtered_exon:\n",
    "        testcount += 1\n",
    "        #print(i)\n",
    "print(testcount) \n",
    "#print((genes_test_df_Framingham_filtered_exon))\n",
    "print(len(genes_test_df_Framingham_filtered_exon))\n",
    "print(len(set(genes_test_df_Framingham_filtered_exon)))\n",
    "\n",
    "hash_exons_Framingham = {}\n",
    "for i in range(0, len(genes_test_df_Framingham_filtered_exon)):\n",
    "    hash_exons_Framingham.setdefault(genes_test_df_Framingham_filtered_exon[i], [])\n",
    "    (hash_exons_Framingham[genes_test_df_Framingham_filtered_exon[i]]).append(rsIDs_test_df_Framingham_filtered_exon[i])\n",
    "print(hash_exons_Framingham.keys())\n",
    "\n",
    "list_of_exons_with_ciseQTL_perturbing_TF_binding = []\n",
    "for k, v in hash_exons_Framingham.items():\n",
    "    for i in rsIDs_with_altered_binding:\n",
    "        if i in v:\n",
    "            list_of_exons_with_ciseQTL_perturbing_TF_binding.append(k)\n",
    "print(set(list_of_exons_with_ciseQTL_perturbing_TF_binding))\n",
    "print(len(set(list_of_exons_with_ciseQTL_perturbing_TF_binding)))\n",
    "for i in list_of_exons_with_ciseQTL_perturbing_TF_binding:\n",
    "    if i not in hash_exons_Framingham.keys():\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### There were 5 types of nans:\n",
    "### Manually searched:\n",
    "# chr 11: 5275530\t5275739\n",
    "# rs11036431\t\tNOT IN rsIDs_with_altered_binding\n",
    "\n",
    "# chr 11: 5275867\t5275903\n",
    "#rs11036635\n",
    "#rs7119142\n",
    "#rs16912979\t\tYES IN\n",
    "#rs72872555\n",
    "#rs11036641\n",
    "\n",
    "# chr 11: 25688127\t25688348\n",
    "#rs669063\t\tNOT IN\n",
    "\n",
    "#chr 13: 31034028\t31034507\n",
    "#rs9578178\n",
    "#rs12431309\n",
    "#rs9508757\n",
    "#rs9579588\n",
    "#rs9578179\t\tYES IN\n",
    "\n",
    "# chr 2: 61343127\t61343182\n",
    "# rs1896293\t\tNOT IN\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "53843\n",
      "204599\n",
      "38654\n",
      "140739\n",
      "38654\n",
      "140739\n"
     ]
    }
   ],
   "source": [
    "### Extracting relevant information regarding fine-mapped putative causal variants from Ulirsch Nat Genet 2019\n",
    "\n",
    "import pandas as pd, os, re\n",
    "\n",
    "hash_all_fine_mapped_var_0001 = {}\n",
    "hash_rsIDs_with_PP_1percent = {}\n",
    "hash_rsIDs_with_PP_0point1 = {}\n",
    "\n",
    "pd_df_finemapped = pd.read_excel(\"41588_2019_362_MOESM3_ESM.xlsx\", sheetname = 0, header = 2)\n",
    "list_identifier = pd_df_finemapped[\"VAR_ID\"].tolist()\n",
    "list_rsID = pd_df_finemapped[\"rsID\"].tolist()\n",
    "list_PP = pd_df_finemapped[\"PP\"].tolist()\n",
    "\n",
    "count_PP_001 = 0\n",
    "count_PP_0001 = 0\n",
    "list_of_var_PP_001 = []\n",
    "list_of_var_PP_0001 = []\n",
    "list_of_rsID_PP_001 = []\n",
    "list_of_rsID_PP_0001 = []\n",
    "\n",
    "for i in range(0, len(list_identifier)):\n",
    "    hash_all_fine_mapped_var_0001[i] = (list_identifier[i], list_rsID[i], list_PP[i])\n",
    "    if float(list_PP[i]) > 0.01:\n",
    "        count_PP_001 += 1\n",
    "        if list_identifier[i] not in list_of_var_PP_001:\n",
    "            list_of_var_PP_001.append(list_identifier[i])\n",
    "            if list_rsID[i]:\n",
    "                list_of_rsID_PP_001.append(list_rsID[i])\n",
    "    if float(list_PP[i]) > 0.001:\n",
    "        count_PP_0001 += 1\n",
    "        if list_identifier[i] not in list_of_var_PP_0001:\n",
    "            list_of_var_PP_0001.append(list_identifier[i])\n",
    "            if list_rsID[i]:\n",
    "                list_of_rsID_PP_0001.append(list_rsID[i])\n",
    "print(count_PP_001)\n",
    "print(count_PP_0001)\n",
    "print(len(list_of_var_PP_001)) ### There are 38654 unique variants that are PP > 0.01\n",
    "print(len(list_of_var_PP_0001)) ### There are 140739 unique variants that are PP > 0.001\n",
    "print(len(list_of_rsID_PP_001))\n",
    "print(len(list_of_rsID_PP_0001))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1557\n",
      "448\n",
      "184\n"
     ]
    }
   ],
   "source": [
    "### Getting list of rsIDs with perturbed TF binding\n",
    "\n",
    "list_STFs = ['Egr1_2580_contig8mers', 'EGR2_REF_R1_8mers', 'Gli1_v015681_8mers_11111111', \n",
    "             'Gli2_v015681_8mers_11111111', 'Gli3_v015681_8mers_11111111', \n",
    "             'HES1_R1', 'HES5_R1', 'HES7_R1', 'Rara_1051_contig8mers', \n",
    "             'Rarb_8mers_11111111', 'Rarg_8mers_11111111', 'Rxra_1035_contig8mers', 'Rxrg_8mers_11111111', \n",
    "             'Smad3_3805_contig8mers', 'TCF4_R1']\n",
    "\n",
    "list_GATA = ['Gata3_1024_contig8mers', 'Gata4_8mers_11111111', 'Gata5_3768_contig8mers', 'Gata6_3769_contig8mers', 'GATA_average']\n",
    "\n",
    "rsIDs_with_altered_STF_binding = []\n",
    "for i in list_rsIDs_PBMs_of_interest:\n",
    "    if i[0] not in rsIDs_with_altered_STF_binding:\n",
    "        if i[1] in list_STFs:\n",
    "            rsIDs_with_altered_STF_binding.append(i[0])\n",
    "print(len(rsIDs_with_altered_STF_binding))\n",
    "\n",
    "### There are 1557 rsIDs_with_altered_STF_binding amongst the 3263 SNPs\n",
    "\n",
    "rsIDs_with_altered_GATA_binding = []\n",
    "for i in list_rsIDs_PBMs_of_interest:\n",
    "    if i[0] not in rsIDs_with_altered_GATA_binding:\n",
    "        if i[1] in list_GATA:\n",
    "            rsIDs_with_altered_GATA_binding.append(i[0])\n",
    "print(len(rsIDs_with_altered_GATA_binding))\n",
    "\n",
    "### There are 448 rsIDs_with_altered_GATA_binding amongst the 3263 SNPs\n",
    "\n",
    "rsIDs_with_altered_GATA_average_binding = []\n",
    "for i in list_rsIDs_PBMs_of_interest:\n",
    "    if i[0] not in rsIDs_with_altered_GATA_average_binding:\n",
    "        if i[1] == \"GATA_average\":\n",
    "            rsIDs_with_altered_GATA_average_binding.append(i[0])\n",
    "print(len(rsIDs_with_altered_GATA_average_binding))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "542\n",
      "911\n",
      "158\n",
      "276\n",
      "62\n",
      "115\n",
      "1064\n",
      "1911\n"
     ]
    }
   ],
   "source": [
    "### Considering intersection between 3263 SNPs and fine-mapped putative causal variants from Ulirsch Nat Genet 2019\n",
    "\n",
    "count_rsIDs_PP_001_altered_STF_binding = 0\n",
    "count_rsIDs_PP_0001_altered_STF_binding = 0\n",
    "count_rsIDs_PP_001_altered_GATA_binding = 0\n",
    "count_rsIDs_PP_0001_altered_GATA_binding = 0\n",
    "count_rsIDs_PP_001_altered_GATA_average_binding = 0\n",
    "count_rsIDs_PP_0001_altered_GATA_average_binding = 0\n",
    "count_rsIDs_PP_001_all = 0\n",
    "count_rsIDs_PP_0001_all = 0\n",
    "\n",
    "#file_to_write_rsIDs_PP_001_all = open(\"rsIDs_PP_001_all.txt\", \"w\")\n",
    "#file_to_write_rsIDs_PP_0001_all = open(\"rsIDs_PP_0001_all.txt\", \"w\")\n",
    "\n",
    "for i in rsIDs_with_altered_STF_binding:\n",
    "    if i in list_of_rsID_PP_001:\n",
    "        count_rsIDs_PP_001_altered_STF_binding += 1\n",
    "    if i in list_of_rsID_PP_0001:\n",
    "        count_rsIDs_PP_0001_altered_STF_binding += 1\n",
    "\n",
    "for i in rsIDs_with_altered_GATA_binding:\n",
    "    if i in list_of_rsID_PP_001:\n",
    "        count_rsIDs_PP_001_altered_GATA_binding += 1\n",
    "    if i in list_of_rsID_PP_0001:\n",
    "        count_rsIDs_PP_0001_altered_GATA_binding += 1\n",
    "\n",
    "for i in rsIDs_with_altered_GATA_average_binding:\n",
    "    if i in list_of_rsID_PP_001:\n",
    "        count_rsIDs_PP_001_altered_GATA_average_binding += 1\n",
    "    if i in list_of_rsID_PP_0001:\n",
    "        count_rsIDs_PP_0001_altered_GATA_average_binding += 1\n",
    "\n",
    "for j in hash_RBC_SNPs_sliding_window.keys():\n",
    "    if j in list_of_rsID_PP_001:\n",
    "        count_rsIDs_PP_001_all += 1\n",
    "#        file_to_write_rsIDs_PP_001_all.write(j + \"\\n\")\n",
    "    if j in list_of_rsID_PP_0001:\n",
    "        count_rsIDs_PP_0001_all += 1\n",
    "#        file_to_write_rsIDs_PP_0001_all.write(j + \"\\n\")\n",
    "\n",
    "#file_to_write_rsIDs_PP_001_all.close()\n",
    "#file_to_write_rsIDs_PP_0001_all.close()\n",
    "print(count_rsIDs_PP_001_altered_STF_binding) # 542 for STFs, PP > 0.01\n",
    "print(count_rsIDs_PP_0001_altered_STF_binding) # 911 for STFs, PP > 0.001\n",
    "print(count_rsIDs_PP_001_altered_GATA_binding) # 158 for GATA, PP > 0.01\n",
    "print(count_rsIDs_PP_0001_altered_GATA_binding) # 276 for GATA, PP > 0.001\n",
    "print(count_rsIDs_PP_001_altered_GATA_average_binding) # 62 for GATA_average, PP > 0.01\n",
    "print(count_rsIDs_PP_0001_altered_GATA_average_binding) # 115 for GATA_average, PP > 0.001\n",
    "print(count_rsIDs_PP_001_all) # 1064 for all 3263 rsIDs\n",
    "print(count_rsIDs_PP_0001_all) # 1911 for all 3263 rsIDs\n",
    "\n",
    "### These values can be used to perform Fisher's exact tests (two-sided) to look at intersection of\n",
    "### the 3263 single nucleotide substitution SNPs in this study and fine-mapped blood trait SNPs."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:py36]",
   "language": "python",
   "name": "conda-env-py36-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
